tsdb/chunks/chunks.go

```diff
@@ -252,63 +254,144 @@ func MergeOverlappingChunks(chks []Meta) ([]Meta, error) {
 			last++
 			continue
 		}
-		nc := &newChks[last]
-		if c.MaxTime > nc.MaxTime {
-			nc.MaxTime = c.MaxTime
-		}
-		chk, err := MergeChunks(nc.Chunk, c.Chunk)
+		dividedChunkOne, dividedChunkTwo, err := MergeChunks(newChks[last].Chunk, c.Chunk, MaxSamplesInChunk)
 		if err != nil {
 			return nil, err
 		}
-		nc.Chunk = chk
+		mint, maxt := getMinAndMaxTimeOfChunk(dividedChunkOne)
+		newChks[last] = Meta{Chunk: dividedChunkOne, MinTime: mint, MaxTime: maxt}
+		if dividedChunkTwo != nil {
+			mint, maxt = getMinAndMaxTimeOfChunk(dividedChunkTwo)
+			newMeta := Meta{Chunk: dividedChunkTwo, MinTime: mint, MaxTime: maxt}
+			newChks = append(newChks, newMeta)
+			last++
+		}
 	}
-
 	return newChks, nil
 }
 
 // MergeChunks vertically merges a and b, i.e., if there is any sample
 // with same timestamp in both a and b, the sample in a is discarded.
-func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
+// Also breaks chunks that are larger than a particular limit.
+func MergeChunks(a, b chunkenc.Chunk, limit int) (chunkenc.Chunk, chunkenc.Chunk, error) {
```

Comment:
Would it be possible to return two `Meta`s instead of the Chunks? It seems like you could maybe use the times that are being iterated over in here to get the mint/maxt and that would avoid the use of the `getMinAndMaxTimeOfChunk` fucntion

Reply:
Great suggestion. I think the maintainers will like it too.